File: 1b.hs
{-# LANGUAGE OverloadedStrings, LambdaCase #-}
module Main where

import Data.List(group)
import qualified Data.Set as S

main = do
  text <- readFile "1.txt"
  let solution = solve $ map readNum $ lines text
  putStrLn $ show solution

solve xs =
  let frequencies = scanl (+) 0 $ cycle xs
      foundSoFar = scanl (\found next -> S.insert next found) S.empty frequencies
      withPreviouslyFound = tail frequencies `zip` foundSoFar
      repeated = map fst $ filter (uncurry S.member) withPreviouslyFound
  in head repeated

readNum :: String -> Int
readNum ('+' : rest) = readNum rest
readNum str = read str

File: 1b.js
const fs = require("fs")

const directions = fs.readFileSync("1.txt", "UTF-8")
  .split("\n")
  .map(x => parseInt(x))
  .filter(x => !Number.isNaN(x))

const solve = (directions) => {
  const found = {}
  let i = 0
  let current = 0
  while (true) {
    let next = directions[(i++) % directions.length] + current
    if (found[next]) {
      return next
    }
    found[next] = true
    current = next
  }
}

console.log(solve(directions))
